### PR TITLE
Hello C# - Migrating the codebase from C++/CX to C# (Phase 1)

### PR DESCRIPTION
**Pull request by tian-lt**
## Hello C# - Migrating the codebase from C++/CX to C# (Phase 1)
This PR is Related to issue #893

### Description of the changes:
- Migrates the View layer from C++/CX to C#
- Changes CalcViewModel into a Windows Runtime Component library
- Adjusts pipelines and project configuration for the C# codebase

#### Manual Tests Results for Branch [feature/UICSharpCalculator](https://github.com/microsoft/calculator/tree/feature/UICSharpCalculator)
- **Test State**: Completed
- **Test Health**: Green
- **Platforms(s)**: Desktop VM & Desktop & Surface Pro 7+ & HP laptop 
- **OS Build(s)**: RS4 (17134.1.180410-1804), RS5 (17763.1.180914-1434), 21H1(19043.1052), VB (19041.1.191206-1406), Co (22000.1)
- **App Build**: Calculator -Release 0.2106.1402.0

### CODE REVIEW

**Comment by tian-lt** [comment]
Hi everyone,
It's so glad to proposing this PR to bring the UI part of Calculator codebase to C# world.
Though this PR is changing too many files, we've set up a feature branch([feature/UICSharpCalculator](https://github.com/microsoft/calculator/tree/feature/UICSharpCalculator)) to track and break down changes. So, for each important change, there should be a corresponding PR that helps you focus on key parts while reviewing the whole picture.
Feel free to leave your comments here.
Appreciate it.

File: build/pipelines/azure-pipelines.ci.yaml [hunk 1]
```diff
@@ -5,11 +5,11 @@
 #
 
 trigger:
-- master
```
Comment (mcooley) [inline build/pipelines/azure-pipelines.ci.yaml:8]: Revert this change and keep `master`.

File: src/CalcViewModel/CalcViewModel.vcxproj.filters [hunk 2]
```diff
@@ -1,28 +1,25 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup>
+    <Filter Include="Resources">
```
Comment (mcooley) [inline src/CalcViewModel/CalcViewModel.vcxproj.filters:4]: nit: the "Resources" filter can be removed

File: src/Calculator/WindowFrameService.cs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        }
+
+        public void RegisterOnWindowClosingHandler(Action onWindowClosingHandler)
+        {
+            m_onWindowClosingHandlers.Add(onWindowClosingHandler);
+        }
+
+        public Task HandleViewRelease()
+        {
+            var that = this;
```
Comment (mcooley) [inline src/Calculator/WindowFrameService.cs:59]: nit: there's no longer a reason to use `that`. `this` can be used in the lambda below.

File: src/Calculator/Views/UnitConverter.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+        }
+
+        private void InitializeOfflineStatusTextBlock()
+        {
+            var resProvider = AppResourceProvider.GetInstance();
+            string offlineStatusHyperlinkText = resProvider.GetResourceString("OfflineStatusHyperlinkText");
+
+            // The resource string has the 'NetworkSettings' hyperlink wrapped with '%HL%'.
+            // Break the string and assign pieces appropriately.
+            const string delimiter = "%HL%";
```
Comment (mcooley) [inline src/Calculator/Views/UnitConverter.xaml.cs:307]: nit: we should consider using string.Split to simplify this code

File: src/Calculator/Calculator.csproj [hunk 5]
```diff
@@ -0,0 +1,848 @@
...
+    <AppxPackageSigningTimestampDigestAlgorithm>SHA256</AppxPackageSigningTimestampDigestAlgorithm>
+    <AppxAutoIncrementPackageRevision>False</AppxAutoIncrementPackageRevision>
+    <GenerateTestArtifacts>True</GenerateTestArtifacts>
+    <AppxBundlePlatforms Condition="'$(Platform)' == 'x64'">x64</AppxBundlePlatforms>
+    <AppxBundlePlatforms Condition="'$(Platform)' == 'x86'">x86</AppxBundlePlatforms>
+    <AppxBundlePlatforms Condition="'$(Platform)' == 'ARM'">ARM</AppxBundlePlatforms>
+    <AppxBundlePlatforms Condition="'$(Platform)' == 'ARM64'">ARM64</AppxBundlePlatforms>
+    <HoursBetweenUpdateChecks>0</HoursBetweenUpdateChecks>
+  </PropertyGroup>
+  <!-- This has to be exactly in this place for this to work -->
```
Comment (mcooley) [inline src/Calculator/Calculator.csproj:42]: nit: PreferredToolArchitecture is probably no longer necessary and this can be removed

File: src/Calculator/Calculator.csproj [hunk 6]
```diff
@@ -0,0 +1,848 @@
...
+    <DefineConstants>$(DefineConstants);SEND_DIAGNOSTICS;IS_STORE_BUILD</DefineConstants>
+  </PropertyGroup>
+  <PropertyGroup>
+    <RestoreProjectStyle>PackageReference</RestoreProjectStyle>
+  </PropertyGroup>
+  <PropertyGroup>
+    <AppVersion Condition="'$(AppVersion)' == ''">0.0.0.0</AppVersion>
+  </PropertyGroup>
+  <ItemGroup>
+      <ResourceCompile Include="Calculator.rc" PreprocessorDefinitions="%(PreprocessorDefinitions);APP_VERSION_MAJOR=$(AppVersion.Split(`.`)[0]);APP_VERSION_MINOR=$(AppVersion.Split(`.`)[1]);APP_VERSION_BUILD=$(AppVersion.Split(`.`)[2]);APP_VERSION_REVISION=$(AppVersion.Split(`.`)[3])" />
```
Comment (mcooley) [inline src/Calculator/Calculator.csproj:149]: The .rc file can be removed. The approach I've used in other projects is to include this .targets file to generate a .cs file with version information:
```
<?xml version="1.0" encoding="utf-8" ?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">

  <!--
    For .NET Core SDK-style projects, the "Version" MSBuild property is written into assembly files
    automatically. This target provides the same feature in non-SDK-style projects using a
    generated file.
  -->

  <PropertyGroup Condition="'$(Version)' != ''">
    <CompileDependsOn>StampAssemblyInfo;$(CompileDependsOn)</CompileDependsOn>
  </PropertyGroup>

  <Target Name="StampAssemblyInfo">
    <ItemGroup>
      <AssemblyAttributes Include="AssemblyVersion">
        <_Parameter1>$(Version)</_Parameter1>
      </AssemblyAttributes>
    </ItemGroup>
    <WriteCodeFragment AssemblyAttributes="@(AssemblyAttributes)"
                        Language="C#"
                        OutputDirectory="$(IntermediateOutputPath)"
                        OutputFile="GeneratedAssemblyInfo.cs">
        <Output TaskParameter="OutputFile" ItemName="Compile" />
        <Output TaskParameter="OutputFile" ItemName="FileWrites" />
    </WriteCodeFragment>
   </Target>

</Project>
```

**Review by mcooley — COMMENTED** [review]
This is looking very good! I didn't review the whole thing, but I checked a few of the build-related files.

File: build/pipelines/azure-pipelines.release.yaml [hunk 7]
```diff
@@ -16,8 +16,8 @@ pr: none
 
 variables:
   versionMajor: 10
-  versionMinor: 2103
-  versionBuild: $[counter('10.2103.*', 0)]
+  versionMinor: 2108
```
Comment (tian-lt) [inline build/pipelines/azure-pipelines.release.yaml:19]: Hi @mcooley, is it okay to bump the version to 2018 even though we're going to release it before August?

File: build/pipelines/azure-pipelines.ci.yaml (hunk 1 above)
Comment (tian-lt) [reply build/pipelines/azure-pipelines.ci.yaml]: done

File: src/CalcViewModel/CalcViewModel.vcxproj.filters (hunk 2 above)
Comment (tian-lt) [reply src/CalcViewModel/CalcViewModel.vcxproj.filters]: done - removed from CalcViewModel and CalcViewModelCopyForUT

File: src/Calculator/WindowFrameService.cs (hunk 3 above)
Comment (tian-lt) [reply src/Calculator/WindowFrameService.cs]: Good point - done

File: src/Calculator/Views/UnitConverter.xaml.cs (hunk 4 above)
Comment (tian-lt) [reply src/Calculator/Views/UnitConverter.xaml.cs:307]: nice suggestion, we'll improve this part later after the C# work gets finished.

File: src/Calculator/Calculator.csproj (hunk 5 above)
Comment (tian-lt) [reply src/Calculator/Calculator.csproj:39]: done

File: src/Calculator/Calculator.csproj (hunk 6 above)
Comment (tian-lt) [reply src/Calculator/Calculator.csproj]: Glad to see we have such interesting solution. Fixed in the commit [8915ce1](https://github.com/microsoft/calculator/pull/1598/commits/8915ce11057ab17366acacb71acd7949dad993b6). Plz let me know if I did it wrong. Thx.

File: build/pipelines/azure-pipelines.release.yaml (hunk 7 above)
Comment (mcooley) [reply build/pipelines/azure-pipelines.release.yaml:19]: This is OK.

File: src/build/Calculator.StampAssemblyInfo.targets [hunk 8]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <!--
+    For .NET Core SDK-style projects, the "Version" MSBuild property is written into assembly files
+    automatically. This target provides the same feature in non-SDK-style projects using a
+    generated file.
+  -->
+
+  <PropertyGroup Condition="'$(Version)' != ''">
```
Comment (mcooley) [inline src/build/Calculator.StampAssemblyInfo.targets:10]: You'll probably want to use `$(AppVersion)` instead of `$(Version)` in this file, since AppVersion is what's being passed on the command line [here](https://github.com/microsoft/calculator/blob/233339d289ca1f422b3371e99a20311f93625ade/build/pipelines/templates/build-single-architecture.yaml#L32).

Another option would be to change build-single-architecture.yaml to use `Version` instead of `AppVersion`. If you do that you'll need to change other projects in the repo which currently use `AppVersion`.

File: src/Calculator/Calculator.csproj (hunk 6 above)
Comment (mcooley) [reply src/Calculator/Calculator.csproj]: New changes mostly look good to me--see my other comment in the .targets file for the one issue I noticed.

**Review by hanzhang54 — DISMISSED** [review]
Thank you Matt and Tian for reviewing and updating. This looks very good to me!

File: src/build/Calculator.StampAssemblyInfo.targets (hunk 8 above)
Comment (tian-lt) [reply src/build/Calculator.StampAssemblyInfo.targets:10]: Thanks Matt,
I decided to use `Version` instead of `AppVersion` since build-single-architecture.yaml is the only producer who specifies `AppVersion`. Please see commit [43634b9](https://github.com/microsoft/calculator/pull/1598/commits/43634b9c368d038ed4112ab7d935889a1a825e81) for details.

Also, edited AssemblyInfo.cs to provide the information that was declared in Calculator.rc. Changes are in commit [
4ed33a7](https://github.com/microsoft/calculator/pull/1598/commits/4ed33a7d405edf314889bcd25c5be62ae482d527).

After CI-pipeline run finished, I checked the properties of file CalculatorApp.exe and it has the expected information.
image

Comment (mcooley) [reply src/build/Calculator.StampAssemblyInfo.targets:10]: Sounds good--just make sure that the .dll files in the package still have version information just like CalculatorApp.exe.

Comment (tian-lt) [reply src/build/Calculator.StampAssemblyInfo.targets:10]: Double checked the master branch, only Calculator project(the UI project) has assembly-info. The other projects, such as ***CalcViewModel***, ***GraphControl***, ***GraphingImpl***, ***TraceLogging***, they are tagging with default assembly-info which is empty. ***CalcViewModel*** was a static-library, so it's fine if it doesn't have assembly-info, but the remaining projects don't look good.

I've assigned some assembly-info for all the shared library(.dll) projects in commit [8926e99](https://github.com/microsoft/calculator/pull/1598/commits/8926e99d7efbaf22557c4065534cb91fcc2f5b72). Let me know if it's not good to do so and I'll revert the commit.

BTW, if we want to add assembly-info for ***GraphingImpl.dll***, we need to submit some changes in the internal repo. But that task should not block this PR from getting checked in, I suppose.

**Review by hanzhang54 — APPROVED** [review]
LGTM

**Comment by xuhongxu96** [comment]
`DesignData` has not been converted?

File: src/Calculator/Views/StateTriggers/CalculatorProgrammerDisplayPanel.xaml [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<UserControl x:Class="CalculatorApp.CalculatorProgrammerDisplayPanel"
+﻿<UserControl x:Class="CalculatorApp.CalculatorProgrammerDisplayPanel"
```
Comment (xuhongxu96) [inline src/Calculator/Views/StateTriggers/CalculatorProgrammerDisplayPanel.xaml:1]: Why move `CalculatorProgrammerDisplayPanel` into `StateTriggers` folder?

**Review by xuhongxu96 — CHANGES_REQUESTED** [review]

File: src/Calculator/App.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+                                            safeFrameServiceCreation.SetOperationSuccess(true);
+                                        }
+                                        else
+                                        {
+                                            var activatedEventArgs = (args as IApplicationViewActivatedEventArgs);
+                                            if ((activatedEventArgs != null) && (activatedEventArgs.CurrentlyShownApplicationViewId != 0))
+                                            {
+                                                // CSHARP_MIGRATION_ANNOTATION:
+                                                // here we don't use ContinueWith() to interpret origin code because we would like to 
+                                                // pursue the design of class SafeFrameWindowCreate whichi was using RAII to ensure
```
[the remark this answers is not in the document]
Comment (tian-lt) [reply src/Calculator/App.xaml.cs:266]: thanks, let's make it correct in other PR.

File: src/Calculator/Views/StateTriggers/CalculatorProgrammerDisplayPanel.xaml (hunk 9 above)
Comment (tian-lt) [reply src/Calculator/Views/StateTriggers/CalculatorProgrammerDisplayPanel.xaml:1]: `CalculatorProgrammerDisplayPanel` was putting under the wrong place - `StateTriggers` - in the master branch. There're several problems similar to this. Since we don't want to pause this PR for a long time, let's fix it up later.

File: src/build/Calculator.StampAssemblyInfo.targets (hunk 8 above)
Comment (tian-lt) [reply src/build/Calculator.StampAssemblyInfo.targets:10]: Hi @mcooley , this PR has been paused for a while. We'd like to make it checked in first. If there's any problems, we can send out other PRs to fix them.

**Comment by tian-lt** [comment]
> `DesignData` has not been converted to C#?

The `DesignData` files are not being included as translation units. Keeping it as what it is doesn't block the progress of C# migration.
